Framework/MDAlgorithms/src/IntegrateEllipsoidsTwoStep.cpp

```diff
@@ -0,0 +1,553 @@
+#include "MantidMDAlgorithms/IntegrateEllipsoidsTwoStep.h"
+
+#include "MantidAPI/DetectorInfo.h"
+#include "MantidAPI/InstrumentValidator.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/Sample.h"
+#include "MantidDataObjects/PeaksWorkspace.h"
+#include "MantidDataObjects/PeakShapeEllipsoid.h"
+#include "MantidDataObjects/Workspace2D.h"
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidGeometry/Crystal/IndexingUtils.h"
+#include "MantidGeometry/Crystal/OrientedLattice.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/CompositeValidator.h"
+#include "MantidKernel/make_unique.h"
+#include "MantidKernel/NearestNeighbours.h"
+#include "MantidMDAlgorithms/Integrate3DEvents.h"
+#include "MantidMDAlgorithms/MDTransfFactory.h"
+#include "MantidMDAlgorithms/MDTransfQ3D.h"
+#include "MantidMDAlgorithms/UnitsConversionHelper.h"
+
+#include <cmath>
+#include <string>
+#include <tuple>
+#include <boost/math/special_functions/round.hpp>
+
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Kernel;
+
+namespace Mantid {
+namespace MDAlgorithms {
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(IntegrateEllipsoidsTwoStep)
+
+//---------------------------------------------------------------------
+/// Algorithm's name for identification. @see Algorithm::name
+const std::string IntegrateEllipsoidsTwoStep::name() const {
+  return "IntegrateEllipsoidsTwoStep";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int IntegrateEllipsoidsTwoStep::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string IntegrateEllipsoidsTwoStep::category() const {
+  return "Crystal\\Integration";
+}
+
+void IntegrateEllipsoidsTwoStep::init() {
+  auto ws_valid = boost::make_shared<CompositeValidator>();
+  ws_valid->add<InstrumentValidator>();
+
+  auto mustBePositive = boost::make_shared<BoundedValidator<double>>();
+  mustBePositive->setLower(0.0);
+
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input, ws_valid),
+                  "An input MatrixWorkspace with time-of-flight units along "
+                  "X-axis and defined instrument with defined sample");
+
+  declareProperty(make_unique<WorkspaceProperty<PeaksWorkspace>>(
+                      "PeaksWorkspace", "", Direction::InOut),
+                  "Workspace with peaks to be integrated");
+
+  declareProperty("RegionRadius", .35, mustBePositive,
+                  "Only events at most this distance from a peak will be "
```

Comment:
Copy and pasted from IntegrateEllipsoids. These properties should also be more explicit about what units they are using.